### PR TITLE
Allow keyword list in context functions

### LINKED ISSUE #125: `Sentry.Context.add_breadcumb` should accept keywords list instead/alongside map

### PR DESCRIPTION
**Pull request by hauleth**
Closes #125

### CODE REVIEW

File: lib/sentry/context.ex [hunk 1]
```diff
@@ -1,78 +1,74 @@
...
+  def get_all do
+    context = get_context()
 
-  def set_tags_context(map) when is_map(map) do
-    get_context()
-    |> set_context(@tags_key, map)
+    for {_, key} <- @keys, into: %{} do
+      {key, Map.get(context, key, %{})}
+    end
+    |> Map.put(@breadcrumbs_key, Map.get(context, :breadcrumbs, []))
```
Comment (hauleth) [inline lib/sentry/context.ex]: TBD add `breadcrumbs` to `@keys` and simplify this to just comprehension.

File: lib/sentry/context.ex [hunk 2]
```diff
@@ -1,78 +1,74 @@
...
-    new_context = Map.put(current_context, @breadcrumbs_key, new_breadcrumbs)
-    Process.put(@process_dictionary_key, new_context)
+  @spec add_breadcrumb(data) :: any()
+  def add_breadcrumb(map) do
+    map = map
+          |> Map.new
+          |> Map.put_new(:timestamp, Sentry.Util.unix_timestamp())
+
+    context = get_context()
+              |> Map.update(@breadcrumbs_key, [map], &(&1 ++ [map]))
```
Comment (hauleth) [inline lib/sentry/context.ex]: Review if Sentry API requires sorted list, if not then `[map | &1]` should be faster than `&1 ++ [map]`.

File: lib/sentry/context.ex [hunk 3]
```diff
@@ -1,78 +1,81 @@
...
+          |> Map.new
+          |> Map.put_new(:timestamp, Sentry.Util.unix_timestamp())
+
+    context = get_context()
+              |> Map.update(@breadcrumbs_key, [map], &([map | &1]))
+
+    Process.put(@process_dictionary_key, context)
+  end
+
+  def pop_breadcrumb do
```
Comment (hauleth) [inline lib/sentry/context.ex]: I am not quite sure if this method should exist but I have added it as it was so simple.

File: lib/sentry/context.ex [hunk 4]
```diff
@@ -1,78 +1,81 @@
...
   end
 
-  def set_http_context(map) when is_map(map) do
-    get_context()
-    |> set_context(@request_key, map)
+  Enum.each @keys, fn {name, key} ->
+    @spec unquote(:"set_#{name}_context")(data):: any()
+    def unquote(:"set_#{name}_context")(map) do
+      map
+      |> Map.new
```
Comment (hauleth) [inline lib/sentry/context.ex]: I wanted to add 3rd element in tuples that would be `default` and use `Enum.into(unquote default)` but hit the wall with `%{}`. I need to investigate it more.

File: lib/sentry/context.ex [hunk 5]
```diff
@@ -1,78 +1,81 @@
...
+  def get_all do
+    context = get_context()
 
-  def set_tags_context(map) when is_map(map) do
-    get_context()
-    |> set_context(@tags_key, map)
+    for {_, key} <- @keys, into: %{} do
+      {key, Map.get(context, key, %{})}
+    end
+    |> Map.put(@breadcrumbs_key, Enum.reverse(Map.get(context, :breadcrumbs, [])))
```
Comment (hauleth) [inline lib/sentry/context.ex]: Maybe this whole structure should be a `defstruct` and `Sentry.Context.get_all/0` should be removed in favor of `Sentry.Context.get_context/0` (which I believe, could be renamed simply to `Sentry.Context.get/0` then).

**Comment by mitchellhenke** [comment]
Sorry for the delay in responding.  I appreciate the time you put into this 🙂 

Could you reduce the changes to just those around accepting keywords or maps for now?

**Comment by hauleth** [comment]
Closing in favour of #139

**Pull request closed without merging**